File: apps/mobile_android/lib/trusted_contacts.dart
/// Trusted contacts — list of designated people who get notified if
/// a run goes wrong (overdue finish, panic button, etc).
///
/// Persona-hunt Round 3 finding Woman #4. Mobile twin of
/// `apps/web/src/lib/trusted_contacts.ts`. Keep both in lockstep —
/// the `shared-library-syncer` agent flags divergence.
///
/// Pure-data + pure-functions module. No Flutter / Supabase deps.

class TrustedContact {
  final String name;
  final String? phone;
  final String? email;
  final String? relationship;

  const TrustedContact({
    required this.name,
    this.phone,
    this.email,
    this.relationship,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        if (phone != null) 'phone': phone,
        if (email != null) 'email': email,
        if (relationship != null) 'relationship': relationship,
      };

  factory TrustedContact.fromJson(Map<String, dynamic> j) => TrustedContact(
        name: (j['name'] as String?) ?? '',
        phone: j['phone'] as String?,
        email: j['email'] as String?,
        relationship: j['relationship'] as String?,
      );
}

const String kTrustedContactsKey = 'trusted_contacts';

/// Cap pinned to the web side. Surface to UI rather than failing
/// silently when the user tries to add a sixth.
const int kMaxTrustedContacts = 5;

/// Normalise a single contact — trim every string field, drop empties.
/// Returns null when `name` is missing (name is required; an entry
/// without it is not a contact).
TrustedContact? normaliseTrustedContact(TrustedContact input) {
  final name = input.name.trim();
  if (name.isEmpty) return null;
  final phone = input.phone?.trim();
  final email = input.email?.trim();
  final relationship = input.relationship?.trim();
  return TrustedContact(
    name: name,
    phone: (phone != null && phone.isNotEmpty) ? phone : null,
    email: (email != null && email.isNotEmpty) ? email : null,
    relationship:
        (relationship != null && relationship.isNotEmpty) ? relationship : null,
  );
}

/// Normalise an entire list — drops invalid entries + caps at
/// kMaxTrustedContacts. Returns a fresh list; never mutates input.
List<TrustedContact> normaliseTrustedContacts(List<TrustedContact>? input) {
  if (input == null || input.isEmpty) return const [];
  final out = <TrustedContact>[];
  for (final c in input) {
    final n = normaliseTrustedContact(c);
    if (n != null) out.add(n);
    if (out.length >= kMaxTrustedContacts) break;
  }
  return out;
}

/// True when the contact has at least one reachable channel.
bool hasReachableChannel(TrustedContact c) {
  return (c.phone != null && c.phone!.isNotEmpty) ||
      (c.email != null && c.email!.isNotEmpty);
}

File: apps/web/src/lib/social/trusted_contacts.ts
// Trusted contacts — list of designated people who get notified if
// a run goes wrong (overdue finish, panic button, etc).
//
// Persona-hunt Round 3 finding Woman #4. A runner heading out for a
// solo long run wants to designate one or more trusted contacts as
// part of their pre-run safety routine. This module is the scaffold:
// the data type + pref key + a single sanitiser. The actual
// notify-on-overdue / panic-button delivery logic is deferred until
// the safety surface lands — we want the storage shape locked in
// first so runners can start populating their list ahead of the
// feature.
//
// Pure functions, no Svelte / Supabase dependencies — same shape as
// `privacy.ts`. Mobile twin lives at
// `apps/mobile_android/lib/trusted_contacts.dart` (byte-identical to
// the iOS twin per the one-Dart-codebase rule).

export interface TrustedContact {
	/// Required. Display name of the contact. Trimmed; empty rejected
	/// at the editor.
	name: string;
	/// E.164 phone number (or local-format — we don't validate yet).
	/// Optional because some users will prefer email-only.
	phone?: string;
	/// Email address. Same optionality as phone.
	email?: string;
	/// Free-form relationship label ("partner", "parent", "running
	/// buddy"). Optional. Surface-only — never required.
	relationship?: string;
}

export const TRUSTED_CONTACTS_KEY = 'trusted_contacts';

/// Reasonable cap so a runaway editor can't blow up the universal-
/// prefs bag. Five is enough to cover spouse + parent + best friend +
/// run partner + emergency operator without becoming an address book.
export const MAX_TRUSTED_CONTACTS = 5;

/// Normalise a single contact: trim every string field, drop empties.
/// Returns null if `name` is missing — name is the only required
/// field and an entry without it is not a contact.
export function normaliseTrustedContact(input: TrustedContact): TrustedContact | null {
	const name = input.name?.trim() ?? '';
	if (name.length === 0) return null;
	const out: TrustedContact = { name };
	const phone = input.phone?.trim();
	if (phone) out.phone = phone;
	const email = input.email?.trim();
	if (email) out.email = email;
	const relationship = input.relationship?.trim();
	if (relationship) out.relationship = relationship;
	return out;
}

/// Normalise an entire list — drops invalid entries + caps at
/// MAX_TRUSTED_CONTACTS. Returns a fresh array; never mutates input.
export function normaliseTrustedContacts(
	input: TrustedContact[] | null | undefined,
): TrustedContact[] {
	if (!input) return [];
	const out: TrustedContact[] = [];
	for (const c of input) {
		const n = normaliseTrustedContact(c);
		if (n) out.push(n);
		if (out.length >= MAX_TRUSTED_CONTACTS) break;
	}
	return out;
}

/// Quick safety check at editor-submit time: returns true when the
/// contact has at least one reachable channel. The notify-on-overdue
/// path (when it lands) is useless without a phone OR email — a
/// name-only entry would silently produce zero notifications. UI
/// surfaces a warning rather than blocking, since a runner may still
/// want to record someone they'll call manually.
export function hasReachableChannel(c: TrustedContact): boolean {
	return Boolean((c.phone && c.phone.length > 0) || (c.email && c.email.length > 0));
}
